Add a sample code for reading an input matrix market file for a simple cuDSS resolution

rbourgeois33 opened the pull request:
This MR adds a sample code for reading an input matrix market file for a simple cuDSS resolution. If the user gives the provided matrices, it will be the exact same resolution as the one in the original "simple" example.

Let me know if this is useless (is there a reader somewhere else ?). If useful, let me know what I can do to improve if necessary.

rbourgeois33 commented:
Note: I just realized that the "simple" example code assumes that the input matrix is symmetric, and that we only feed the upper triangular part to the solver.

How to change this ? I would like to give a non-symmetric matrix to cuDSS

thanks !

kvoronin commented:
Thanks for the PR, we'll review this shortly! I definitely see the value of such a sample, it will be quite useful.

rbourgeois33 commented:
Hi @kvoronin . Happy to help!

 I added a few functionalities to the reader / resolution script that I found helpful when trying out cuDSS on our CFD matrices:
 
1.  Added timers to the different parts of the resolution
2. Compute the final error, taking into account the view type
3. implemented a easy to use command line input
`Usage: ./binary <alg: 0|1|2|3> <mtype: general|symmetric|hermitian|spd|hpd> <mview: full|lower|upper> <matrix_filename> [vector_filename (optional)]
`
4. Use a vector filled with 1 if no rhs is given
5. Add warnings if algs 1 or 2 are used with a mtype that is not general
6. Display a warning if the mview is not 'full' while mtype is 'general'
7. Throw a runtime error if the reader finds elements in the upper / lower part of the matrix and the mtype is lower / upper
8. Throw a runtime error if the reader does not find elements in both upper / lower parts of the matrix and the mview is full.

Hope this helps,
Rémi

essex-edwards commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,127 @@
+            lineData >> i >> j >> val;
+
+            if (i > previous_line) {
+                (*csr_offsets_h)[i_offsets++] = cumulative_nnz;
```
It looks like this line requires that the entries in the MTX file are sorted by row. I don't believe that's a requirement of the MTX file format. Nor is that true of, e.g. the MTX files from the SuiteSparse collection. IMHO, this limitation is acceptable, but I would suggesting adding some code to detect if the input is not sorted that way and printing an explanatory error message in that case:
```
else if(i < previous_line) { std::cerr<<"...." ... }
```

It also looks like this line requires that all rows will be non-empty. That will be true of useful inputs to cuDSS, but not of MTX files in general. It would probably also be good to check that too.

rbourgeois33 commented:
Hello,

Thanks again for detecting the missing corner cases of my mtx reader. Some update:

-We now accept unsorted entries, and will sort them with std::sort, as well as empty rows.
-The reader does a small mock allocation if no file is found so that subsequent free operations don't fail.
-Throws an error if nnz in the header is != to the # of elements in the mtx file
-Throws an error is read i,j are out of bound i.e <0, or >= n
-Display a warning for empty rows (maybe its a bit too much)
-add a verbose mode for infos and errors

Rémi

kvoronin commented:
Hi @rbourgeois33!

Sorry for the long wait, but thanks to this PR, we have re-visited our policy for accepting external contributions and now implemented the changes which allow us to accept your contribution. 

We have changed the license of the repository from BSD-3 to Apache 2.0 to allow accepting external contributions in the right way.

What this means is that
- the new default branch is `main`. (The old one called `master` is still there, but `master` and an extra `bsd3_main` are not supposed to be changed)
- all commits to the repository must be signed, see the new https://github.com/NVIDIA/CUDALibrarySamples/blob/main/CONTRIBUTING.md. Technically, it is as simple as doing `git commit -s ...` which would enforce DCO for the contribution (see more details in the CONTRIBUTING.md).

What this means for this PR:
Please change the target branch to `main` (potentially instead of rebase it is easier to cherry-pick commits into a new branch) and sign the commits.

Thanks!
